### ISSUE TITLE
Support length on splat or count interpolation

### ISSUE THREAD

**Issue by mengesb** [hashicorp/terraform#7718]
I'd like to see support for `${length(aws_instance.named_instance.*.id)}` whereby the length reported is the count of `aws_instance.named_instance._.id` elements occur.

Logically I'd think that `${length(split(",",join(",",aws_instance.named_instance.*.id)))}` would do the trick, however I'm getting errors with this syntax and it looks rather messy. I'm not sure if it is related to me using `${length(keys(var.aws_subnets))}` as my count for that instance construct versus an integer, however I would like to use a length/count on the splat.

Either `length()` supporting splats or adding another interpolation `count()` would be handy in this circumstance.

**Comment by cornfeedhobo** [comment]
Happening to me in 0.7.0-rc3. This would be great to see in 0.7!

**Comment by carlsverre** [comment]
:+1: this would be super useful!

**Comment by apparentlymart** [comment]
I think I remember support being added for interpolation expressions like `aws_instance.named_instance.count`. Perhaps it would be simplest to make *that* be allowed in `count`, if it isn't already, since that way we avoid special cases around how `length` evaluates for splats, which doesn't really fit into the current evaluator design.

**Comment by mengesb** [comment]
Unfortunately that means they have to be `integers` I think...

```hcl
> aws_instance.vormetric.count
Error reading aws_instance.vormetric count: strconv.ParseInt: parsing "${var.host_count[\"vormetric\"]}": invalid syntax
```

**Comment by apparentlymart** [comment]
I think that may just be a bug in how that attribute gets interpolated (or not, as the case may be) when in `console`, which I think is what you used there. Other resource attributes come directly from state attributes of the same name but that one is a bit of a special case that might not behave right in this context due to the special way it is implemented.

I'm not at my computer right now to dig into the code but I think it might work if used actually in config, which is a more "normal" environment for interpolations.

**Comment by mengesb** [comment]
Hard to trust that when console is around and emits that

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence here.

In the master branch now we have a change to make `length` work in this case. The root change here is to allow for functions that are able to accept a list that might contain unknown values, whereas before there was a hard rule that any unknown values would automatically short-circuit the function creation and produce an unknown result.

Another aspect of those changes is that an expression like `aws_instance.named_instance` now returns a list of objects, so you can actually say `length(aws_instance.named_instance)` and get the same result, without the need for a splat expression.

Since these changes are already in master, I'm going to close this out. The changes will be included in the forthcoming v0.12.0 release.
